File: 1028-recover-a-tree-from-preorder-traversal/1028-recover-a-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int i = 0;
    public TreeNode recoverTree(String s,int depth){
      int d= 0;
        while(i+d<s.length() && s.charAt(i+d) == '-'){
            d++;
        }
        if(d!=depth)return null;
        int nd = 0;
        while(i+d+nd<s.length() && s.charAt(i+d+nd)!='-'){
            nd++;
        }
        int val = Integer.parseInt(s.substring(i+d,i+d+nd));
        i = i+d+nd;
        TreeNode root = new TreeNode(val);
        root.left = recoverTree(s,depth+1);
        root.right = recoverTree(s,depth+1);
     return root;
        
    }
    public TreeNode recoverFromPreorder(String traversal) {
       
        TreeNode root = recoverTree(traversal,0);
        return root;
    }
}

File: 1028-recover-a-tree-from-preorder-traversal/1028-recover-a-tree-from-preorder-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode RecoverFromPreorder(string S) {
        Dictionary<int, TreeNode> _lastAtLevel = new Dictionary<int, TreeNode>();
        
        int currentLevel=0;
        int lastIndex=0;
        
        while(lastIndex<S.Length){
            currentLevel=0;
            
            while(lastIndex<S.Length && S[lastIndex]=='-'){
                lastIndex++;
                currentLevel++;
            }

            int valStartIndex = lastIndex;

            while(lastIndex<S.Length && S[lastIndex]!='-'){
                lastIndex++;
            }

            string val=S.Substring(valStartIndex, lastIndex-valStartIndex);

            TreeNode tn = new TreeNode(int.Parse(val));

            if(currentLevel>0){
                TreeNode parent = _lastAtLevel[currentLevel-1];

                if(parent.left==null){
                    parent.left = tn;
                }
                else{
                    parent.right = tn;
                }
            }

            _lastAtLevel[currentLevel]=tn;
        }
        
        return _lastAtLevel[0];
    }   
}